xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-requiredrights/xwiki-platform-security-requiredrights-default/src/main/java/org/xwiki/platform/security/requiredrights/internal/XWikiDocumentRequiredRightAnalyzer.java

```diff
@@ -0,0 +1,111 @@
+import org.xwiki.platform.security.requiredrights.RequiredRightAnalyzer;
+import org.xwiki.platform.security.requiredrights.RequiredRightsException;
+import org.xwiki.rendering.block.XDOM;
+import org.xwiki.security.authorization.Right;
+
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * @version $Id$
+ * @since 15.8RC1
+ */
+@Component
+@Singleton
+@Named(XWikiDocumentRequiredRightAnalyzer.ID)
+public class XWikiDocumentRequiredRightAnalyzer implements RequiredRightAnalyzer<XWikiDocument>
+{
+    /**
+     * The id of this component.
+     */
+    public static final String ID = "document";
+
+    @Inject
+    private DocumentContextExecutor documentContextExecutor;
+
+    @Inject
+    @Named("translation")
+    private BlockSupplierProvider<String> translationMessageSupplierProvider;
+
+    @Inject
+    @Named(XDOMRequiredRightAnalyzer.ID)
+    private RequiredRightAnalyzer<XDOM> xdomRequiredRightAnalyzer;
+
+    @Inject
+    @Named(DefaultObjectRequiredRightAnalyzer.ID)
+    private RequiredRightAnalyzer<BaseObject> objectRequiredRightAnalyzer;
+
+    @Override
+    public List<RequiredRightAnalysisResult> analyze(XWikiDocument document) throws RequiredRightsException
+    {
+        // Analyze the content
+        try {
+            // Push the document into the context such that we, e.g., get the correct context wiki with the correct
+            // wiki macros etc.
+            return this.documentContextExecutor.call(() ->
+            {
+                List<RequiredRightAnalysisResult> result = new ArrayList<>();
+
+                // Analyze the title
+                if (VelocityUtil.containsVelocityScript(document.getTitle())) {
+                    result.add(new RequiredRightAnalysisResult(
+                        document.getDocumentReferenceWithLocale(),
+                        this.translationMessageSupplierProvider.get("security.requiredrights.title"),
+                        this.translationMessageSupplierProvider.get("security.requiredrights.title.description",
+                            document.getTitle()),
+                        List.of(
+                            new RequiredRight(Right.SCRIPT, EntityType.DOCUMENT, true),
+                            new RequiredRight(Right.PROGRAM, EntityType.DOCUMENT, true)
+                        )
+                    ));
+                }
+
+                // Analyze the content
+                result.addAll(this.xdomRequiredRightAnalyzer.analyze(document.getXDOM()));
+
+                for (List<BaseObject> baseObjects : document.getXObjects().values()) {
+                    for (BaseObject object : baseObjects) {
+                        result.addAll(this.objectRequiredRightAnalyzer.analyze(object));
+                    }
+                }
+
+                return result;
+            }, document);
+        } catch (Exception e) {
+            throw new RequiredRightsException("Error...", e);
```

Comment:
Missing a proper message.

Reply:
I added a comment on https://github.com/xwiki/xwiki-platform/pull/2470/files#r1363467928 on the topic. It looks good to me to have a defined exception and that each analyzer might throw it. Now I think by default when traversing the doc you should catch those and not stop at first analyzer error: it would be a pity that if adding a custom buggy analyzer it would break the whole analysis.